### ISSUE TITLE
Searching in hierarchical data structures

### ISSUE THREAD

**Issue by kz282** [biagioT/jpa-search-helper#10]
Hello Biagio,
I am trying your jpa-search-helper with a hierarchical data structure, and I always get java.lang.StackOverflowError at the first findAll method call. If I remove the NestedSearchable annotation, it is not possible to traverse the hierarchy.

By checking your sources I found that the issue is with ReflectionUtils.getFields(). I was thinking if it was possible to collect info for those fields that are necessary for the query, and not trying to process all of them. Or collect the fields by the Entity classes and do not process them more than once.

```
@Data
@Entity
@ToString
@Accessors(chain=true)
public class Ruler {
    
    @Id
    @GeneratedValue
    private Long id;
    
    @Searchable
    private String name;

    @ManyToOne(optional = true)
    @NestedSearchable
    public Ruler predecessor;
    
}
```

```
public class RulerTest {
    private static final Logger logger = LogManager.getLogger();
    
    @Autowired
    private RulerRepository rulerRepository;
    
    @Test
    @Transactional
    public void rulerTest(){
        try{
            Ruler Iulius = rulerRepository.save(new Ruler().setName("Iulius Caesar"));
            Ruler Augustus = rulerRepository.save(new Ruler().setName("Augustus").setPredecessor(Iulius));

            logger.info("Iulius: {}", Iulius);
            logger.info("Augustus: {}", Augustus);

            List<Ruler> findAugustus = rulerRepository.findAll(Map.of("name_eq", "Augustus"), Ruler.class);
            logger.info("Find Augustus: {}", findAugustus);

            List<Ruler> findAugustus2 = rulerRepository.findAll(Map.of("predecessor.name_eq", "Iulius Caesar"), Ruler.class);
            logger.info("Find Augustus by predecessor: {}", findAugustus2);
        }catch(Throwable th){
            logger.error(th.getMessage(), th);
            fail(th);
        }
    }
}
```

**Comment by biagioT** [comment]
Hi!
thanks for the report. The bug was fixed in [3.3.1](https://github.com/biagioT/jpa-search-helper/releases/tag/3.3.1) version, available now on Maven Central.

Let me know if it's ok! If you like, add a star to the project :smile:

**Comment by kz282** [comment]
I have checked the new release, and Augustus is now found. :D Thanks for your quick fix!
